File: src/main/java/org/kite9/diagram/visualization/orthogonalization/edge/EdgeConverter.java
package org.kite9.diagram.visualization.orthogonalization.edge;

import java.util.List;
import java.util.Map;

import org.kite9.diagram.common.elements.edge.PlanarizationEdge;
import org.kite9.diagram.common.elements.vertex.Vertex;
import org.kite9.diagram.model.DiagramElement;
import org.kite9.diagram.model.position.Direction;
import org.kite9.diagram.visualization.orthogonalization.Dart;
import org.kite9.diagram.visualization.orthogonalization.DartFace.DartDirection;
import org.kite9.diagram.visualization.orthogonalization.Orthogonalization;

public interface EdgeConverter {

	IncidentDart convertPlanarizationEdge(PlanarizationEdge e, Orthogonalization o, Direction incident, Vertex externalVertex, Vertex sideVertex, Vertex planVertex, Direction fanStep);
	
	void convertContainerEdge(Map<DiagramElement, Direction> underlyings, Orthogonalization o, Vertex end1, Vertex end2, Direction d, List<Dart> s);

	void createEdgePart(Orthogonalization o, Direction direction, Vertex start, Vertex end, DiagramElement forDe, Direction forDeSide, List<DartDirection> out);

	
}

File: src/main/kotlin/org/kite9/diagram/visualization/orthogonalization/edge/EdgeConverter.kt
package org.kite9.diagram.visualization.orthogonalization.edge

import org.kite9.diagram.common.elements.edge.PlanarizationEdge
import org.kite9.diagram.common.elements.vertex.Vertex
import org.kite9.diagram.model.DiagramElement
import org.kite9.diagram.model.position.Direction
import org.kite9.diagram.visualization.orthogonalization.Dart
import org.kite9.diagram.visualization.orthogonalization.Orthogonalization

/**
 * Converts edges into darts.  In subclasses, takes account of fanning and labelling too.
 * @author robmoffat
 */
interface EdgeConverter {

    /**
     * This is used for converting planarization edges (i.e. elements in the planarization) into darts.
     *
     */
    fun convertPlanarizationEdge(
        e: PlanarizationEdge,
        o: Orthogonalization,
        incident: Direction,
        externalVertex: Vertex,
        sideVertex: Vertex,
        planVertex: Vertex,
        fanStep: Direction?
    ): IncidentDart

    /**
     * This is used for creating darts to represent the 2d shape of a vertex, which was a point in the planarization.
     */
    fun buildDartsBetweenVertices(
        underlyings: Map<DiagramElement, Direction?>,
        o: Orthogonalization,
        end1: Vertex,
        end2: Vertex,
        d: Direction
    ): List<Dart>
}